src/version.h

```diff
@@ -0,0 +1,142 @@
+public:
+  enum class ParseMode
+  {
+    // official semver parsing
+    //
+    SemVer,
+
+    // MO2 parsing, e.g., 2.5.1rc1 - this either parse a string with no pre-release
+    // information (e.g. 2.5.1) or with a single pre-release + a version (e.g., 2.5.1a1
+    // or 2.5.2rc1)
+    //
+    MO2
+  };
+
+  enum class ReleaseType
+  {
+    Development,       // -dev
+    Alpha,             // -alpha, -a
+    Beta,              // -beta, -b
+    ReleaseCandidate,  // -rc
+  };
+  using enum ReleaseType;
+
+public:  // parsing
+  // parse version from the given string, throw InvalidVersionException if the string
+  // cannot be parsed
+  //
+  static Version parse(QString const& value, ParseMode mode = ParseMode::SemVer);
+
+public:  // constructor
+  Version(int major, int minor, int patch, QString metadata = {});
+  Version(int major, int minor, int patch, ReleaseType type, QString metadata = {});
+  Version(int major, int minor, int patch, ReleaseType type, int prerelease,
+          QString metadata = {});
+  Version(int major, int minor, int patch,
+          std::vector<std::variant<int, ReleaseType>> prereleases,
+          QString metadata = {});
+
+public:  // special member functions
+  Version(const Version&) = default;
+  Version(Version&&)      = default;
+
+  Version& operator=(const Version&) = default;
+  Version& operator=(Version&&)      = default;
+
+public:
+  // check if this version corresponds to a pre-release version (dev, alpha, beta, etc.)
+  //
+  bool isPreRelease() const { return !m_PreReleases.empty(); }
+
+  // retrieve major, minor and patch of this version
+  //
+  int major() const { return m_Major; }
+  int minor() const { return m_Minor; }
+  int patch() const { return m_Patch; }
+
+  // retrieve pre-releases information for this version
+  //
+  const auto& preReleases() const { return m_PreReleases; }
+
+  // retrieve build metadata, if any, otherwise return an empty string
+  //
+  const auto& buildMetadata() const { return m_BuildMetadata; }
+
+  // convert this version to a semver string
+  //
+  QString string() const;
+
+private:
+  // major.minor.patch
+  int m_Major, m_Minor, m_Patch;
+
+  // pre-release information
+  std::vector<std::variant<int, ReleaseType>> m_PreReleases;
+
+  // metadata
+  QString m_BuildMetadata;
+};
+
+QDLLEXPORT std::strong_ordering operator<=>(const Version& lhs, const Version& rhs);
+
+inline bool operator==(const Version& lhs, const Version& rhs)
+{
+  return (lhs <=> rhs) == 0;
+}
+inline bool operator!=(const Version& lhs, const Version& rhs)
+{
+  return (lhs <=> rhs) != 0;
+}
+inline bool operator<(const Version& lhs, const Version& rhs)
+{
+  return (lhs <=> rhs) < 0;
+}
+inline bool operator>(const Version& lhs, const Version& rhs)
+{
+  return (lhs <=> rhs) > 0;
+}
+inline bool operator<=(const Version& lhs, const Version& rhs)
+{
+  return (lhs <=> rhs) <= 0;
+}
+inline bool operator>=(const Version& lhs, const Version& rhs)
+{
+  return (lhs <=> rhs) >= 0;
+}
```

Comment:
If everything's working properly, these should all be generated by the compiler if there's a spaceship operator, so I think they're either redundant, or indicative of some kind of problem that should ideally be fixed.

Reply:
Yes, I thought that too, but after trying for basically 30 minutes, I did not manage to get the right declaration of `operator<=>` to get these generated so I gave up. If you know the right declaration to use, I'll change it.

Reply:
What was failing to find them? I think they're generated on demand at the callsite. The declaration should be something along the lines of `friend std::strong_ordering operator<=>(const Version&, const Version&) = default;`

Reply:
So apparently this is only true if `<=>` is defaulted, which is not the case here, so the extra definition are needed - https://stackoverflow.com/a/58780963/2666289.

Reply:
That covers `operator==` and `operator!=` not being generated. cppreference says
> The inequality operator is automatically generated by the compiler if `operator==` is defined. Likewise, the four relational operators are automatically generated by the compiler if the three-way comparison operator `operator<=>` is defined. `operator==` and `operator!=`, in turn, are generated by the compiler if `operator<=>` is defined as defaulted

So you only need `operator==` and `operator<=>`.

Reply:
You're right, I misread the cppreference quote (I did not think there would be a different behavior for equality and comparison operators actually). I have removed the extra operators.